internal/pkg/keeper/client.go

```diff
@@ -0,0 +1,213 @@
+func (k *keeperClient) IsAlive() bool {
+	if _, err := k.commonClient.Ping(context.Background()); err != nil {
+		return false
+	}
+	return true
+}
+
+// Register registers the current service with Keeper for discovery and health check
+func (k *keeperClient) Register() error {
+	if k.serviceKey == "" || k.serviceHost == "" || k.servicePort == 0 ||
+		k.healthCheckRoute == "" || k.healthCheckInterval == "" {
+		return fmt.Errorf("unable to register service with keeper: Service information not set")
+	}
+
+	registrationReq := requests.AddRegistrationRequest{
+		BaseRequest: dtoCommon.BaseRequest{
+			Versionable: dtoCommon.Versionable{ApiVersion: common.ApiVersion},
+		},
+		Registration: dtos.Registration{
+			ServiceId: k.serviceKey,
+			Host:      k.serviceHost,
+			Port:      k.servicePort,
+			HealthCheck: dtos.HealthCheck{
+				Interval: k.healthCheckInterval,
+				Path:     k.healthCheckRoute,
+				Type:     "http",
+			},
+		},
+	}
+
+	// check if the service registry exists first
+	resp, err := k.registryClient.RegistrationByServiceId(context.Background(), k.serviceKey)
+	if err != nil && resp.StatusCode != http.StatusNotFound {
+		return fmt.Errorf("failed to check the %s service registry status: %v", k.serviceKey, err)
+	}
+
+	// call the UpdateRegister to update the registry if the service already exists
+	// otherwise, call Register to create a new registry
+	if resp.StatusCode == http.StatusOK {
+		err := k.registryClient.UpdateRegister(context.Background(), registrationReq)
+		if err != nil {
+			return fmt.Errorf("failed to update the %s service registry: %v", k.serviceKey, err)
+		}
+	} else {
+		err := k.registryClient.Register(context.Background(), registrationReq)
+		if err != nil {
+			return fmt.Errorf("failed to register the %s service: %v", k.serviceKey, err)
+		}
+	}
+
+	return nil
+}
+
+// RegisterCheck registers a health check with Keeper
+func (k *keeperClient) RegisterCheck(id string, name string, notes string, url string, interval string) error {
+	// keeper combines service discovery and health check into one single register request
+	return nil
+}
+
+func (k *keeperClient) UnregisterCheck(id string) error {
+	// keeper combines service discovery and health check into one single register request
+	return nil
+}
+
+// Unregister de-registers the current service from Keeper
+func (k *keeperClient) Unregister() error {
+	registrationReq := requests.AddRegistrationRequest{
+		BaseRequest: dtoCommon.BaseRequest{
+			Versionable: dtoCommon.Versionable{ApiVersion: common.ApiVersion},
+		},
+		Registration: dtos.Registration{
+			ServiceId: k.serviceKey,
+			Host:      k.serviceHost,
+			Port:      k.servicePort,
+			HealthCheck: dtos.HealthCheck{
+				Interval: k.healthCheckInterval,
+				Path:     k.healthCheckRoute,
+				Type:     "http",
+			},
+			Status: models.Halt,
+		},
+	}
+
+	err := k.registryClient.UpdateRegister(context.Background(), registrationReq)
```

Comment:
It's an intentional design so that users can still get the information of the deregistered services and do whatever they want to do (e.g. use those information to restart the stopped service)

For the [GetAllServiceEndpoints](https://github.com/edgexfoundry/go-mod-registry/pull/138/files#diff-1284d0c429d2b38d42b7baa22190bec31c0e657e9132f5167f47b8a038a0f6a2R172-R173) method, it will filter out those deregistered services.
But the `GET /registry/all` [API](https://github.com/edgexfoundry-holding/edgex-core-keeper/pull/10/files#diff-0c91d203920341a0a19ed3a997f9b37271d08d5396c4145d7104b4b5d1d24051R65-R92) retains a flexibility to include a boolean parameter named `deregistered`, which determines whether to retrieve services that have been deregistered.